substratevm/src/com.oracle.svm.core.jdk9/src/com/oracle/svm/core/jdk/Target_java_lang_StackWalker.java

```diff
@@ -25,19 +25,340 @@
+    StackWalker.StackFrame next;
+
+    VirtualFrameIterator(final DeoptimizedFrame.VirtualFrame frame) {
+        this.frame = frame;
+    }
+
+    public boolean hasNext() {
+        if (next == null) {
+            if (frame == null) return false;
+            next = new StackFrameImpl(frame.getFrameInfo());
+            frame = frame.getCaller();
+        }
+        return true;
+    }
+
+    public StackWalker.StackFrame next() {
+        if (! hasNext()) throw new NoSuchElementException();
+        try {
+            return next;
+        } finally {
+            next = null;
+        }
+    }
+}
+
+final class FrameInfoIterator implements Iterator<StackWalker.StackFrame> {
+
+    FrameInfoQueryResult frameInfo;
+    StackWalker.StackFrame next;
+
+    FrameInfoIterator(final FrameInfoQueryResult frameInfo) {
+        this.frameInfo = frameInfo;
+    }
+
+    public boolean hasNext() {
+        if (next == null) {
+            if (frameInfo == null) return false;
+            next = new StackFrameImpl(frameInfo);
+            frameInfo = frameInfo.getCaller();
+        }
+        return true;
+    }
+
+    public StackWalker.StackFrame next() {
+        if (! hasNext()) throw new NoSuchElementException();
+        try {
+            return next;
+        } finally {
+            next = null;
+        }
     }
 }
+
+final class StackWalkerHolder {
+    static final Target_java_lang_StackWalker PLAIN = new Target_java_lang_StackWalker(EnumSet.noneOf(StackWalker.Option.class));
+    static final Target_java_lang_StackWalker SRF = new Target_java_lang_StackWalker(EnumSet.of(StackWalker.Option.SHOW_REFLECT_FRAMES));
+    static final Target_java_lang_StackWalker SHF = new Target_java_lang_StackWalker(EnumSet.of(StackWalker.Option.SHOW_HIDDEN_FRAMES));
+}
+
+final class StackFrameImpl implements StackWalker.StackFrame {
+    private final FrameInfoQueryResult frameInfo;
+    private StackTraceElement ste;
+
+    StackFrameImpl(final FrameInfoQueryResult frameInfo) {
+        this.frameInfo = frameInfo;
+    }
+
+    public String getClassName() {
+        final String scn = frameInfo.getSourceClassName();
+        return scn == null ? "" : scn;
+    }
+
+    public String getMethodName() {
+        return frameInfo.getSourceMethodName();
+    }
+
+    public Class<?> getDeclaringClass() {
+        throw new UnsupportedOperationException();
+    }
+
+    public int getByteCodeIndex() {
+        return frameInfo.getBci();
+    }
+
+    public String getFileName() {
+        return frameInfo.getSourceFileName();
+    }
+
+    public int getLineNumber() {
+        return frameInfo.getSourceLineNumber();
+    }
+
+    public boolean isNativeMethod() {
+        return frameInfo.isNativeMethod();
+    }
+
+    public StackTraceElement toStackTraceElement() {
+        StackTraceElement ste = this.ste;
+        if (ste == null) {
+            this.ste = ste = frameInfo.getSourceReference();
+        }
+        return ste;
+    }
```

Comment:
Is there an explanation of this dance with a local `ste` versus the instance `this.ste`?

Is there more here than caching the result of `frameInfo.getSourceReference()`?  Why not just cache it in the instance field?  Atomicity does not seem to be an issue.

Reply:
It's a very common pattern for lazy caching.

Reply:
Okay.  It only came to my attention because our style checker complained about the local variable having the same name as the instance field.  I can deal with that.  I just wanted to know if there was something other than lazy caching going on.